File: bestia-client/src/Bnet/Message/Entity/AttackEntityCMSG.cs
using Bnet;
using Godot;

namespace BestiaBehemothClient.Bnet.Message.Entity
{

  /// <summary>
  /// Swings the active entity's basic attack at a target entity. A basic attack has no entry in the
  /// Attack DB, so nothing here names one - casting a skill is ActivateSkillCMSG.
  /// </summary>
  public partial class AttackEntityCMSG : ICMSG
  {
    [Export] public ulong EntityId { get; set; }

    public AttackEntityCMSG()
    {
    }

    public override Envelope ToEnvelope()
    {
      var attackEntityCmsg = new global::Bnet.AttackEntityCMSG
      {
        EntityId = EntityId
      };

      return new Envelope
      {
        AttackEntity = attackEntityCmsg
      };
    }
  }
}

File: zone-server/src/main/kotlin/net/bestia/zone/battle/AttackEntityCMSG.kt
package net.bestia.zone.battle

import net.bestia.bnet.proto.AttackEntityCmsgProto
import net.bestia.zone.message.CMSG
import net.bestia.zone.util.EntityId

/**
 * A swing of the active entity's basic attack at [targetEntityId]. Carries no attack id and no level:
 * a basic attack has no catalogue row, and a skill is cast with [ActivateSkillCMSG].
 */
data class AttackEntityCMSG(
  override val playerId: Long,
  val targetEntityId: EntityId,
) : CMSG {

  companion object {
    fun fromBnet(
      accountId: Long,
      attackEntity: AttackEntityCmsgProto.AttackEntityCMSG
    ): AttackEntityCMSG {
      return AttackEntityCMSG(
        accountId,
        attackEntity.entityId
      )
    }
  }
}
